usr/lib/ep11_stdll/ep11_specific.c

```diff
@@ -1036,6 +1038,408 @@ CK_RV token_specific_set_attrs_for_new_object(STDLL_TokData_t *tokdata,
  *
  ******************************************************************************/
 
+static CK_RV check_ab_supported(CK_KEY_TYPE type) {
+    switch(type) {
+    case CKK_AES:
+    case CKK_DES2:
+    case CKK_DES3:
+    case CKK_GENERIC_SECRET:
+    case CKK_RSA:
+    case CKK_EC:
+    case CKK_DSA:
+    case CKK_DH:
+        return CKR_OK;
+    default:
+        TRACE_ERROR("%s key type not supported for ab: 0x%lx\n", __func__, type);
+        return CKR_TEMPLATE_INCONSISTENT;
+    }
+}
+
+static CK_RV check_ab_pair(CK_ATTRIBUTE_PTR pPublicKeyTemplate,
+                           CK_ULONG ulPublicKeyAttributeCount,
+                           CK_ATTRIBUTE_PTR pPrivateKeyTemplate,
+                           CK_ULONG ulPrivateKeyAttributeCount)
+{
+    CK_RV rc;
+    CK_BBOOL abpub = FALSE, abpriv = FALSE;
+
+    rc = get_bool_attribute_by_type(pPublicKeyTemplate, ulPublicKeyAttributeCount,
+                                    CKA_IBM_ATTRBOUND, &abpub);
+    if (rc != CKR_OK && rc != CKR_TEMPLATE_INCOMPLETE) {
+        TRACE_ERROR("Failed to retrieve CKA_IBM_ATTRBOUND attribute from public template. rc=0x%lx\n", rc);
+        return rc;
+    }
+    rc = get_bool_attribute_by_type(pPrivateKeyTemplate, ulPrivateKeyAttributeCount,
+                                    CKA_IBM_ATTRBOUND, &abpriv);
+    if (rc != CKR_OK && rc != CKR_TEMPLATE_INCOMPLETE) {
+        TRACE_ERROR("Failed to retrieve CKA_IBM_ATTRBOUND attribute from private template. rc=0x%lx\n", rc);
+        return rc;
+    }
+    if (abpub != abpriv) {
+        TRACE_ERROR("Only one of the key pair is attribute bound.\n");
+        return CKR_TEMPLATE_INCONSISTENT;
+    }
+    return CKR_OK;
+}
+
+static CK_RV check_ab_kek_type(TEMPLATE *tmpl, CK_RV errval)
+{
+    CK_KEY_TYPE kektype;
+    CK_OBJECT_CLASS class;
+    CK_RV rc;
+
+    rc = template_attribute_get_ulong(tmpl, CKA_KEY_TYPE, &kektype);
+    if (rc != CKR_OK)
+        return rc;
```

Comment:
If template_attribute_get_ulong fails, it prints an error level trace entry. However, you can't tell from that trace entry where the call comes from. Because of that it might be useful to also print a trace entry here, where template_attribute_get_ulong is called.

Dependent on the way check_ab_kek_type is used it might be intended to not print additional trace entries. So feel free to leave it as is.....

Reply:
Good catch.  There was actually no other trace message printed on the whole part.  Fixed it for both KEY_TYPE and CLASS.